packages/react/scroll-area/src/ScrollArea.tsx

```diff
@@ -186,2234 +80,886 @@ type ScrollAreaPrimitive = Polymorphic.ForwardRefComponent<
+            onSizesChange({
+              content: context.viewport.scrollHeight,
+              viewport: context.viewport.offsetHeight,
+              scrollbar: {
+                size: ref.current.clientHeight,
+                paddingStart: toInt(computedStyle.paddingTop),
+                paddingEnd: toInt(computedStyle.paddingBottom),
+              },
+            });
+          }
         }}
-        onPointerDown={handlePointerDown}
-        onPointerUp={handlePointerUp}
-        onPointerMove={handlePointerMove}
-        onWheel={handleWheel}
       />
-    </ScrollbarProvider>
+    </ScrollbarAxisProvider>
   );
-}) as ScrollAreaScrollbarPrimitive;
+}) as ScrollAreaScrollbarYPrimitive;
 
-type ScrollAreaScrollbarXOwnProps = Polymorphic.OwnProps<typeof Primitive>;
-type ScrollAreaScrollbarXPrimitive = Polymorphic.ForwardRefComponent<
+/* -----------------------------------------------------------------------------------------------*/
+
+type ScrollAreaScrollbarImplOwnProps = Polymorphic.Merge<
+  Polymorphic.OwnProps<typeof Primitive>,
+  {
+    onWheelScroll(event: WheelEvent): void;
+    onDragScroll(pointerPosition: { x: number; y: number }): void;
+    onResize(): void;
+  }
+>;
+
+type ScrollAreaScrollbarImplPrimitive = Polymorphic.ForwardRefComponent<
   Polymorphic.IntrinsicElement<typeof Primitive>,
-  ScrollAreaScrollbarXOwnProps
+  ScrollAreaScrollbarImplOwnProps
 >;
 
-const ScrollAreaScrollbarX = React.forwardRef(function ScrollAreaScrollbarX(props, forwardedRef) {
-  const { domSizes } = useScrollAreaStateContext();
-  return useNativeScrollArea() ? null : (
-    <ScrollAreaScrollbar
-      {...props}
-      ref={forwardedRef}
-      axis="x"
-      name={SCROLLBAR_X_NAME}
-      style={{
-        ...props.style,
-        // @ts-ignore
-        [SCROLL_AREA_CSS_PROPS.scrollbarXSize]: domSizes.scrollbarX.height
-          ? domSizes.scrollbarX.height + 'px'
-          : 0,
-      }}
-    />
-  );
-}) as ScrollAreaScrollbarXPrimitive;
+const ScrollAreaScrollbarImpl = React.forwardRef((props, forwardedRef) => {
+  const { onDragScroll, onWheelScroll, onResize, ...scrollbarProps } = props;
+  const context = useScrollAreaContext(SCROLLBAR_NAME);
+  const [scrollbar, setScrollbar] = React.useState<ScrollbarElement | null>(null);
+  const composeRefs = useComposedRefs(forwardedRef, (node) => setScrollbar(node));
+  const rectRef = React.useRef<ClientRect | null>(null);
+  const prevWebkitUserSelectRef = React.useRef<string>('');
+  const viewport = context.viewport;
+  const handleWheelScroll = useCallbackRef(onWheelScroll);
+  const debounceResize = useDebounceCallback(onResize, 10);
 
-ScrollAreaScrollbarX.displayName = SCROLLBAR_X_NAME;
+  /**
+   * We bind wheel event imperatively so we can switch off passive
+   * mode for document wheel event to allow it to be prevented
+   */
+  React.useEffect(() => {
+    const opts: AddEventListenerOptions = { passive: false };
```

Comment:
Totally agree, that's what I did first but TS complained 🙁  Had to do this to fix it.

![image](https://user-images.githubusercontent.com/175330/117060440-d6d47400-ad18-11eb-8a3f-dbc08d36f016.png)

Reply:
I see! I guess either that, or add a comment above the DRY line to explain that's because of TS, I don't mind which but both would communicate why we did it.

Reply:
I went with `as any` so we have one less line of code 😂